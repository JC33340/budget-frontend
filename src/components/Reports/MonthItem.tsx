import { reportsDataType } from '../../pages/Reports';

type MonthItemType = {
    item: reportsDataType;
};

const MonthItem = ({ item }: MonthItemType) => {
    //convert number date to string
    const date = new Date();
    date.setMonth(item.month);
    date.setFullYear(item.year);
    let categories: [string, number][] = Object.entries(item.categories);
    //sort categories by value
    categories = categories.sort((a, b) => {
        if (a[1] < b[1]) {
            return 1;
        } else {
            return -1;
        }
    });
    return (
        <div
            className="p-2 border-2 border-dark-gray rounded-md w-full whitespace-nowrap"
            data-testid="monthItem-testid"
        >
            <p className="text-center font-semibold underline text-lg">
                {date.toLocaleDateString('en-GB', { month: 'long' })}{' '}
                <span>({item.year})</span>
            </p>
            <p>
                <span className="font-semibold">Total Balance:</span>{' '}
                <span>{item.balance}</span>
            </p>
            <p className="text-green">
                <span className="font-semibold">Income:</span>{' '}
                <span>{item.income}</span>
            </p>
            <p className="text-red">
                <span className="font-semibold">Expense:</span>{' '}
                <span>{item.expense}</span>
            </p>
            <div>
                <p className="font-semibold">Category Breakdown:</p>
                <div className="max-h-[5rem] overflow-y-scroll border-1 border-light-gray px-2 py-1 rounded-md">
                    {categories.length > 0 ? (
                        categories.map((item, i) => {
                            return (
                                <div className="capitalize" key={i}>
                                    {item[0]}: {item[1]}
                                </div>
                            );
                        })
                    ) : (
                        <p className="text-center">No Data</p>
                    )}
                </div>
            </div>
        </div>
    );
};

export default MonthItem;
