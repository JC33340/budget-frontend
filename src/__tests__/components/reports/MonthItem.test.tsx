import { render, screen } from '@testing-library/react';
import '@testing-library/jest-dom';
import MonthItem from '../../../components/Reports/MonthItem';
import type { reportsDataType } from '../../../pages/Reports';

const testObj: reportsDataType = {
    year: 2025,
    month: 11,
    balance: 999,
    income: 39,
    expense: 56,
    categories: { testCat: 790 },
};

test('snapshot test', () => {
    render(<MonthItem item={testObj} />);
    const item = screen.getByTestId('monthItem-testid');
    expect(item).toMatchSnapshot();
});

test('content test', () => {
    render(<MonthItem item={testObj} />);
    const item = screen.getByTestId('monthItem-testid');
    expect(item).toHaveTextContent('2025');
    expect(item).toHaveTextContent('December');
    expect(item).toHaveTextContent('999');
    expect(item).toHaveTextContent('39');
    expect(item).toHaveTextContent('56');
    expect(item).toHaveTextContent('testCat: 790');
});
