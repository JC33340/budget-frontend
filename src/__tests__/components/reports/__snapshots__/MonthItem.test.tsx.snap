// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshot test 1`] = `
<div
  class="p-2 border-2 border-dark-gray rounded-md w-full whitespace-nowrap"
  data-testid="monthItem-testid"
>
  <p
    class="text-center font-semibold underline text-lg"
  >
    December
     
    <span>
      (
      2025
      )
    </span>
  </p>
  <p>
    <span
      class="font-semibold"
    >
      Total Balance:
    </span>
     
    <span>
      999
    </span>
  </p>
  <p
    class="text-green"
  >
    <span
      class="font-semibold"
    >
      Income:
    </span>
     
    <span>
      39
    </span>
  </p>
  <p
    class="text-red"
  >
    <span
      class="font-semibold"
    >
      Expense:
    </span>
     
    <span>
      56
    </span>
  </p>
  <div>
    <p
      class="font-semibold"
    >
      Category Breakdown:
    </p>
    <div
      class="max-h-[5rem] overflow-y-scroll border-1 border-light-gray px-2 py-1 rounded-md"
    >
      <div
        class="capitalize"
      >
        testCat
        : 
        790
      </div>
    </div>
  </div>
</div>
`;
